modules/openapi-generator/src/main/resources/csharp-netcore/libraries/httpclient/ApiClient.mustache

```diff
@@ -189,30 +187,28 @@ namespace {{packageName}}.Client
         /// <summary>
         /// Initializes a new instance of the <see cref="ApiClient" />, defaulting to the global configurations' base url.
         /// </summary>
-        public ApiClient()
+        public ApiClient(HttpClient client = null, HttpClientHandler handler = null) : this({{packageName}}.Client.GlobalConfiguration.Instance.BasePath, client, handler)
         {
-            _baseUrl = {{packageName}}.Client.GlobalConfiguration.Instance.BasePath;
-    {{#reUseHttpClient}}
-            _httpClientHandler = new HttpClientHandler();
-            _httpClient = new HttpClient(_httpClientHandler);
-    {{/reUseHttpClient}}
         }
 
         /// <summary>
-        /// Initializes a new instance of the <see cref="ApiClient" />
+            /// Initializes a new instance of the <see cref="ApiClient" />
         /// </summary>
         /// <param name="basePath">The target service's base path in URL format.</param>
         /// <exception cref="ArgumentException"></exception>
-        public ApiClient(String basePath)
+        public ApiClient(String basePath, HttpClient client = null, HttpClientHandler handler = null)
         {
             if (string.IsNullOrEmpty(basePath))
                 throw new ArgumentException("basePath cannot be empty");
 
             _baseUrl = basePath;
-    {{#reUseHttpClient}}
-            _httpClientHandler = new HttpClientHandler();
-            _httpClient = new HttpClient(_httpClientHandler);
-    {{/reUseHttpClient}}
+            /* TODO: Decide how to handle this case
+            if(client != null && handler == null) {
+                throw new ArgumentException("if providing HttpClient, you also need to provide its handler");
+            }*/
```

Comment:
From the code sample in the readme:
```
HttpClientHandler yourHandler = new HttpClientHandler();
HttpClient yourHttpClient = new HttpClient(yourHandler);
```
Looks like if the HttpClient is provided, one can obtain its HttpClientHandler, right?

If that's the case, I would suggest simply ask for HttpClient in the constructor to make it easier for the user.

Reply:
That is unfortunately not the case. You can not acquire the handler from the HttpClient once it is constructed. The handler is  used for setting/retrieving cookies as well as configuring the proxy and client certificates. We could say that in case a user passed in their own HttpClient that they would be handling all of that themselves or simply have to live with the fact that it doesn't work. Hence why I was wondering whether we should at least give some kind of warning.

Reply:
Just saw this: https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclient?view=net-5.0#httpclient-and-net-core

Shall we simply use `HttpClient = new HttpClient();` as it should pick the best available transport based on the platform? In other words, ApiClient will not store the HttpClientHandler.

Reply:
Again, we are no longer able to set proxy options and client certificates anymore if we do that. Instantiating HttpClientHandler creates an underlying actual handler which is SocketsHttpHandler.

Reply:
Just to be clear, what we are doing by default is exactly what HttpClient and its default constructor are doing:
        public HttpClient()
            : this(new HttpClientHandler())
        {
        }

Reply:
Having a unique public static instance of HttpClient and HttpClientHandler is not the right way.
You need to inject them from the ctor. You cannot have the lifecycle of each ApiClient bound to this singletons, is not the way is intended to use in net core.

Reply:
We do not have a static instance in this pull request

Reply:
Sorry i reviewed the other one.
It is correct if you inject it, but if it is created inside then is no more disposed.

Reply:
If possible you keep the HttpClient alive for as long as possible, if you create a new one per request even if you dispose of it you will face SocketExhaustion pretty soon. HttpClientFactory would be a good solution its not available for .netstandard 2.0 however only as a nuget package.

You are correct that we should probably dispose the HttpClient when the ApiClient is disposed, if it was created initially. I'll add that.

Reply:
Reading the [guide](https://docs.microsoft.com/en-us/dotnet/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests#issues-with-the-original-httpclient-class-available-in-net)
Here the scenarios:
- Disposing each time the handler is not enough because windows has a tcp timeout that keep connected the socket after the dispose (I can also say that there is also the keep-alive default true to think about), so socket exaustion. So the handler must be pooled, the client must be recreated each times fith the disposeHandler parameter set to off.
- keeping static shared is viable only for some scenarios like small lifed programs (for the DNS issue, if the handler became unhealthy).
So the best way is to:
1. create each times a client, with disposeHandler false, passing the right handler pooled
2. keep an in memory concurrent pool with an evict time to avoid the long timed issue

And this is what the [DefaultHttpClientFactory](https://github.com/aspnet/HttpClientFactory/blob/master/src/Microsoft.Extensions.Http/DefaultHttpClientFactory.cs) does (see also [here](https://docs.microsoft.com/en-us/dotnet/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests#benefits-of-using-ihttpclientfactory)).
Unfortunately the factory has high coupling with the DI packages (logging, options, ...)

So possibilities:
- we code, test, something similar
- we avoid that at all, keeping only the constructor with the injection
- document well the usage of a new HttpClient\Handler not disposed each call, so people is aware of that 

I hope i can help you all to build the netcore client, let me know your thoughts.

Reply:
Actually found another bug while implementing this.

Reply:
We can actually dispose the client if created inside, for sure is better than dont do that at all.
And document the issue in the ctor or in a good place, to let uderstand clearly to the devs that if they don't manage theirself the lifecycle outside, the behavior is to create\dispose.

Reply:
Check the latest commit. It now disposes when the ApiClient or the ApiAccessClass is disposed. If not passed in one HttpClient per ApiAccessClass is created.

Reply:
The commit is ok, it is all right, to bound the httpClient to the entire apiClient lifecycle can emprove multiple calls, and I like it. One thing: if you pass the handler to the client ctor and you dispose both at the end, you need to pass also the second parameter disposeHandler to false. In that way you dispose only the one created inside in all the scenarios.

On the counterpart to moving the ApiClient to an IDisposable in't too much strong? 
The disposable interface will lead to the using scenario and scopes, and this is not useful when injected. 
Another possibility is to keep the lifecycle of the client\hander (if created) inside the ExecAsync call:
`
private async Task<ApiResponse<T>> ExecAsync<T>(HttpRequestMessage req .... )
{
       var httpClientHandler = _httpClientHandler ?? new HttpClientHandler();
       var disposeHandler = _httpClientHandler == null;
       var httpClient = _httpClient ?? new HttpClient(httpClientHandler, false);
       var disposeClient = _httpClient == null;

       ....

       if(disposeClient ) httpClient.Dispose();
       if(disposeHandler) httpClientHandler.Dispose();

       return result;
}
`

Reply:
No, we will not scope the lifecycle to the ExecAsync. We have that in the current version, it leads to socket exhaustion even with disposing, so that is not an option. If you dispose the handler manually if its already disposed nothing will happen, so that is fine to do.

Moving ApiClient to IDisposable is no issue at all.

Reply:
Ok fine, watch out the disposeHandler parameter and all is fine.

Reply:
Yep added that

Reply:
I just realized that I cannot get the handler from the client, and the DI give us just the client yet configured with inside the handler. I will check the NSwag implementation that pass just the client to understand better.

Reply:
Ok NSwag is more simple than that, there is no support for that options related to the handler.
So, if I inject just the client (i think the normal behavior), i loose some settings, this can lead to some misunderstand.
If I inject both, i have the all configuration woriking, but is not so clean in the base DI of a net core project.
I think we can document the settings and the ctor to warn about that.
I also prefer to document some remarks in the ctor to inform in the intellisense about the optional parameters, the behaviors, and the dispose meaning.
Do you think is too much?

Reply:
> I just realized that I cannot get the handler from the client, and the DI give us just the client yet configured with inside the handler. I will check the NSwag implementation that pass just the client to understand better.

Yeah, that limitation is the reason why I have to ask for the handler in the first place. If you inject just the client you will get an exception right now telling you you need to also inject the handler.

Reply:
> Ok NSwag is more simple than that, there is no support for that options related to the handler.
> So, if I inject just the client (i think the normal behavior), i loose some settings, this can lead to some misunderstand.
> If I inject both, i have the all configuration woriking, but is not so clean in the base DI of a net core project.
> I think we can document the settings and the ctor to warn about that.
> I also prefer to document some remarks in the ctor to inform in the intellisense about the optional parameters, the behaviors, and the dispose meaning.
> Do you think is too much?

If you'd like to add further details to the parameters feel free to fork my version of the repo and open a pull request to my branch with the changes.

Reply:
I also thinked about the validation of both parameters as you.
But actually i think that I won't to get a complex DI for these few not so useful parameters.

Reply:
Yeah you need customize the DI injection by running a lambda function that then sets the correct configuration at that point.

Right now adding just the HttpClient is not a supported UseCase. If you say this is relevant for you as well, and you want to use the client from IHttpClientFactory then we'll figure out a way to do that. 

I talked about this also with @wing328 and we actually wanted to add a way to do this later if people request it. It appears you may have a usecase already?

Reply:
I've added a new version that includes a switch to disable the exception and the features that don't work if handler isn't present.

Reply:
I don't understand why in those 3+ years of net core, no one had this issue using a swagger codegen.
Maybe because in the official startup documentation they promote NSwag, effectively the only one actually support the injection of an HttpClient. They also underline very well the issues\tradeoffs in the HttpClient and how they manage it with the factory in the DI.

We moved to net core with a few projects in this years, those services consume others using rest api, with a Swagger codgen. In this days I monitored a strange pressure on the production environment, and I found the socket exsaustion, done by ourself with a chatty client. I compared the different behavior of the vanilla codegen in both framework, monitoring with wireshark:
- one uses 1 socket, 1 TCP handshake, 1 TLS handshale, N calls
- the other one open each times the pipeline consuming sockets

The breaking change is inside RestSharp, it has an issue open but the project seems aware to this difficult change to support net core and I think it has no future plans to fix it. The real change is inside the HttpWebRequest that actually in Net Core no more has a static lifecycle and create a new HttpClient.

So RestSharp is not a valid production library for net core, at all.
This means also Swagger is not a valid choice for net core.
Swagger seems no more active, the main author, William, is here now on OpenApi, reviewing you now (William thank you for the projects, they are great). 

So the choice for us are:
- create new contracts with NSwag for the net core projects, leaving Swagger.
- cooperate on OpenApi to have a net-core client fully integrated in the core DI. OpenApi c# codegen is 99% the same as swagger so moving to OpenApi it seems the right way to continue using Swagger.

Your fix is just enough to avoid the exsaustion now.
But having a client that not fit easily in the DI can move us to choose another more solid way for the future.
But I also see a lot of interest in this PR and on this project to move to a working client for net core so I will await and cooperate with you if I can give some support.

Reply:
I've different suggestions but let's go with what you've so far. We can resume the discussion after the merge.

Reply:
I've different suggestions but let's go with what you've so far. We can resume the discussion after the merge.